packages/migrate/Migration.js

```diff
@@ -0,0 +1,221 @@
+        if (this.config.provider && this.config.provider.engine) {
+          this.config.provider.engine.silent = true;
+        }
+      }
+    } catch (error) {
+      const payload = {
+        type: "migrateErr",
+        error: error,
+      };
+
+      await this.emitter.emit("error", payload);
+      deployer.finish();
+      throw error;
+    }
+  }
+
+  // ------------------------------------- Public -------------------------------------------------
+  /**
+   * Instantiates a deployer, connects this migration and its deployer to the reporter
+   * and launches a migration file's deployment sequence
+   * @param  {Object}   options  config and command-line
+   */
+  async run(options) {
+    const {
+      interfaceAdapter,
+      resolver,
+      context,
+      deployer,
+    } = this.prepareForMigrations(options);
+
+    // Connect reporter to this migration
+    if (this.reporter) {
+      this.reporter.setMigration(this);
+      this.reporter.setDeployer(deployer);
+      this.reporter.confirmations = options.confirmations || 0;
+      this.reporter.listen();
+    }
+
+    // Get file path and emit pre-migration event
+    const file = path.relative(options.migrations_directory, this.file);
+    const block = await interfaceAdapter.getBlock("latest");
+
+    const preMigrationsData = {
+      file: file,
+      number: this.number,
+      isFirst: this.isFirst,
+      network: options.network,
+      networkId: options.network_id,
+      blockLimit: block.gasLimit,
+    };
+
+    await this.emitter.emit("preMigrate", preMigrationsData);
+    await this._load(options, context, deployer, resolver);
+  }
+
+  prepareForMigrations(options) {
+    const logger = options.logger;
+    const interfaceAdapter = createInterfaceAdapter({
+      provider: options.provider,
+      networkType: options.networks[options.network].type,
+    });
+    const web3 = new Web3Shim({
+      provider: options.provider,
+      networkType: options.networks[options.network].type,
+    });
+
+    const resolver = new ResolverIntercept(options.resolver);
+
+    // Initial context.
+    const context = { web3, interfaceAdapter, config: this.config };
+
+    const deployer = new Deployer({
+      logger,
+      confirmations: options.confirmations,
+      timeoutBlocks: options.timeoutBlocks,
+      networks: options.networks,
+      network: options.network,
+      network_id: options.network_id,
+      provider: options.provider,
+      basePath: path.dirname(this.file),
+      ens: options.ens,
+    });
+
+    return { interfaceAdapter, resolver, context, deployer };
+  }
+
+  /**
+   * Returns a serializable version of `this`
+   * @returns  {Object}
+   */
+  serializeable() {
+    return {
+      file: this.file,
+      number: this.number,
+      isFirst: this.isFirst,
+      isLast: this.isLast,
+      dryRun: this.dryRun,
+      interactive: this.interactive,
+    };
+  }
+}
+
+module.exports = Migration;
```

Comment:
re: the new filename for this, I feel like we should come up with a project-wide consistent approach to this sort of thing. In the meantime, I encourage this ad-hoc renaming in the direction of each contributor's personal preference :)

Reply:
I think there is a pretty established convention of naming files containing classes in Pascal case, no? I wouldn't be opposed to meeting about this.

Reply:
yeah I'm not saying anything one way or another about pascal case or whatnot, only that we should at some point make it all consistent

Reply:
you got it dood